// @verida/vda-common/dist/rpc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultRpcUrl = exports.RPC_URLS = void 0;
/* Default RPC URLs to the Polygon blockchains. These default endpoints are public and free but potentially unstable */
exports.RPC_URLS = {
    // Ankr is the official RPC provider of the Polygon foundation
    // https://rpc.ankr.com/polygon is the same as https://polygon-rpc.com as also provided by Ankr
    mainnet: "https://rpc.ankr.com/polygon",
    "0x89": "https://rpc.ankr.com/polygon",
    testnet: "https://rpc-amoy.polygon.technology",
    "0x13882": "https://rpc-amoy.polygon.technology",
    devnet: "https://rpc-amoy.polygon.technology",
    polpos: "https://rpc.ankr.com/polygon",
    polamoy: "https://rpc-amoy.polygon.technology"
};
function getDefaultRpcUrl(chainNameOrId) {
    if (exports.RPC_URLS[chainNameOrId]) {
        return exports.RPC_URLS[chainNameOrId];
    }
    throw new Error("Unknown chain: ".concat(chainNameOrId));
}
exports.getDefaultRpcUrl = getDefaultRpcUrl;
