// @verida/vda-did-resolver/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lookup = exports.activeDIDCount = exports.RPC_URLS = exports.getDIDs = exports.VdaDidResolver = exports.getResolver = void 0;
var resolver_1 = require("./resolver");
Object.defineProperty(exports, "getResolver", { enumerable: true, get: function () { return resolver_1.getResolver; } });
Object.defineProperty(exports, "VdaDidResolver", { enumerable: true, get: function () { return resolver_1.VdaDidResolver; } });
var vda_common_1 = require("@verida/vda-common");
Object.defineProperty(exports, "RPC_URLS", { enumerable: true, get: function () { return vda_common_1.RPC_URLS; } });
var activeDIDCount_1 = require("./activeDIDCount");
Object.defineProperty(exports, "activeDIDCount", { enumerable: true, get: function () { return activeDIDCount_1.activeDIDCount; } });
var getDIDs_1 = require("./getDIDs");
Object.defineProperty(exports, "getDIDs", { enumerable: true, get: function () { return getDIDs_1.getDIDs; } });
var lookup_1 = require("./lookup");
Object.defineProperty(exports, "lookup", { enumerable: true, get: function () { return lookup_1.lookup; } });
