// @verida/types/dist/DocumentInterfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationMethodTypes = exports.SecureContextEndpointType = void 0;
var SecureContextEndpointType;
(function (SecureContextEndpointType) {
    SecureContextEndpointType["DATABASE"] = "database";
    SecureContextEndpointType["MESSAGING"] = "messaging";
    SecureContextEndpointType["STORAGE"] = "storage";
    SecureContextEndpointType["NOTIFICATION"] = "notification";
})(SecureContextEndpointType || (exports.SecureContextEndpointType = SecureContextEndpointType = {}));
var VerificationMethodTypes;
(function (VerificationMethodTypes) {
    VerificationMethodTypes["EcdsaSecp256k1VerificationKey2019"] = "EcdsaSecp256k1VerificationKey2019";
    VerificationMethodTypes["EcdsaSecp256k1RecoveryMethod2020"] = "EcdsaSecp256k1RecoveryMethod2020";
    VerificationMethodTypes["Ed25519VerificationKey2018"] = "Ed25519VerificationKey2018";
    VerificationMethodTypes["RSAVerificationKey2018"] = "RSAVerificationKey2018";
    VerificationMethodTypes["X25519KeyAgreementKey2019"] = "X25519KeyAgreementKey2019";
})(VerificationMethodTypes || (exports.VerificationMethodTypes = VerificationMethodTypes = {}));
